macros/support/src/class/class.rs

```diff
@@ -320,10 +333,111 @@ impl syn::parse::Parse for Class {
 }
 
+/// Resolve name collisions among methods defined on different interfaces, by
+/// renaming some methods with a disambiguating suffix.
+///
+/// It is common for designers of COM interfaces to define multiple versions of
+/// the same interface, where each new interface is a superset of the previous
+/// version. For examples, see these DirectWrite interfaces:
+///
+/// * [IDWriteTextFormat](https://docs.microsoft.com/en-us/windows/win32/api/dwrite/nn-dwrite-idwritetextformat)
+/// * [IDWriteTextFormat1](https://docs.microsoft.com/en-us/windows/win32/api/dwrite_2/nn-dwrite_2-idwritetextformat1)
+/// * [IDWriteTextFormat2](https://docs.microsoft.com/en-us/windows/win32/api/dwrite_3/nn-dwrite_3-idwritetextformat2)
+/// * [IDWriteTextFormat3](https://docs.microsoft.com/en-us/windows/win32/api/dwrite_3/nn-dwrite_3-idwritetextformat3)
+///
+/// These interfaces define methods that have the same name. When implementing a
+/// COM object in Rust, we need a way to distinguish these functions. The ugly
+/// way is to require that different COM interface definitions avoid using the
+/// same name, but that is clearly not feasible.
+///
+/// This implementation looks for name collisions, and simply renames methods
+/// that participate in name collisions by appending the name of the interface.
+/// If _that_ fails, then we use a numeric suffix. (We could just use the IID
+/// as a suffix, but that's a harsh experience when debugging.)
+fn find_method_name_collisions(class: &mut Class) {
+    // Which method names cause collisions?
+    let mut method_name_count: HashMap<Ident, u32> = HashMap::new();
+    for interface_methods in class.methods.values() {
+        for m in interface_methods.iter() {
+            *method_name_count
+                .entry(m.original_ident.clone())
+                .or_default() += 1;
+        }
+    }
+
+    // Scan the methods again. If any of them cause a collision, then create a
+    // new ident for the method definition, using a suffix based on the name of
+    // the interface.  If _that_ fails, fall back to using a numeric suffix.
+    let mut collision_counter: u32 = 0;
+
+    for (interface, methods) in class.methods.iter_mut() {
+        for method in methods.iter_mut() {
+            // We know the unwrap() will succeed, because we're repeating the
+            // same query that we just performed, above.
+            let old_ident = &method.original_ident;
+            let collides = *method_name_count.get(old_ident).unwrap() > 1;
+            if !collides {
+                // Easy, nothing to do.
+                continue;
+            }
+
+            // Rename the method to "bar__IFoo". In most cases, this will work.
+            let interface_ident = path_to_single_ident(interface);
+            let mut new_ident = Ident::new(
+                &format!("{}__{}", interface_ident, old_ident),
+                old_ident.span(),
+            );
+
+            if method_name_count.contains_key(&new_ident) {
```

Comment:
It seems like this would only ever be true in the rare occasion that the user has named a method in the same way as our fallback. I think we should *always* do the method count strategy (and not try to do the `bar__IFoo` strategy first) and always check that the method name we produce is not already taken. 

This would simplify this code and make it easier.